src/quantize_fp16.cpp

```diff
@@ -0,0 +1,165 @@
+        {
+            continue;
+        }
+
+        if(in->get_shape().type() == shape::float_type or
+           in->get_shape().type() == shape::double_type)
+        {
+            assert(in->name() == "convert" and
+                   in->inputs().front()->get_shape().type() == shape::half_type);
+            instruction::replace_argument(ins, in, in->inputs().front());
+        }
+    }
+}
+
+static void quantize_module(module& m,
+                            const std::vector<std::string>& ins_names,
+                            std::unordered_map<instruction_ref, instruction_ref>& map_fp16,
+                            bool quantize_inout = false)
+{
+    for(auto ins : iterator_for(m))
+    {
+        if(ins->name() == "@return")
+        {
+            if(quantize_inout)
+            {
+                convert_outputs_fp16(ins);
+            }
+            break;
+        }
+
+        if(ins->name() == "@param" or ins->name() == "@literal")
+        {
+            auto s = ins->get_shape();
+            if(s.type() == shape::float_type or s.type() == shape::double_type)
+            {
+                auto fp16_ins = m.insert_instruction(
+                    std::next(ins), make_op("convert", {{"target_type", shape::half_type}}), ins);
+                map_fp16[ins] = fp16_ins;
+            }
+
+            continue;
+        }
+
+        if(ins->name() == "convert" and ins->get_shape().type() == shape::half_type)
+        {
+            continue;
+        }
+
+        // all indicates every instruction is converted
+        if((not contains(ins_names, "all")) and (not contains(ins_names, ins->name())))
+        {
+            continue;
+        }
+
+        shape orig_shape = ins->get_shape();
+        // process all inputs, if input is a fp32 or fp64, convert it
+        // to a fp16 by adding a convert operator.
+        auto inputs = ins->inputs();
+        std::vector<instruction_ref> converted_inputs;
+        for(auto input : inputs)
+        {
+            auto s = input->get_shape();
+            if(s.type() != shape::float_type and s.type() != shape::double_type)
+            {
+                converted_inputs.push_back(input);
+                continue;
+            }
+
+            // if the input is a parameter of a subgraph
+            instruction_ref input_fp16{};
+            if(input->name() == "@param")
+            {
+                auto in_outs = input->outputs();
+                auto it      = std::find_if(in_outs.begin(), in_outs.end(), [](auto o) {
+                    return (o->name() == "convert" and o->get_shape().type() == shape::half_type);
+                });
+                assert(it != in_outs.end());
+                input_fp16 = *it;
+                converted_inputs.push_back(input_fp16);
+            }
+            // if the input is a convert operator, uses its input
+            // as its current input
+            else if(input->name() == "convert" and
```

Comment:
There is no need to check if the input is a `convert`, it should always insert a convert and then simplify_reshapes can remove the redundant ones.

Reply:
fixed